### DISCUSSION TITLE
db-scheduler shutting down while some jobs haven't processed

### DISCUSSION THREAD

**Discussion by yzAlvin** [kagkarlsson/db-scheduler#387] [Q&A] [answered]
I am using db-scheduler to queue up one-time tasks, that write to a db.

It seems like my tests are racing with db-scheduler.
* if I add a long `Thread.sleep` then, tests will pass.
* Otherwise the test runner shuts down, and if I insepct the database, `scheduled-tasks` has a bunch of jobs with `picked = false`

Is it possible to configure db-scheduler to keep the test runner alive until all jobs finish processing?
It should delay being shutdown until `scheduled_tasks` table is empty

I understand this might not even be a problem with db-scheduler, and I might need to look at configuring the test instead

**Comment by kagkarlsson** [comment, marked as answer]
In some of my tests I am using the `ManualScheduler` (made for tests) which has the property you require. See [DeleteUnresolvedTest.java](https://github.com/kagkarlsson/db-scheduler/blob/master/db-scheduler/src/test/java/com/github/kagkarlsson/scheduler/functional/DeleteUnresolvedTest.java)

It has a special `ExecutorService` that runs executions in the current thread. I think the method `scheduler.runAnyDueExecutions();`  should run and wait for any scheduled executions, so that asserts can be made after that.

**Reply by yzAlvin** [reply]
Thank you 😊
